src/App.tsx

```
@@ -1,61 +1,126 @@
+}
+
+const todos = todosFromServer.map(todo => {
+  return {
+    ...todo,
+    user: getUserById(todo.userId),
+  };
+});
 
 export const App = () => {
+  const [visibleTodos, setVisibleTodos] = useState(todos);
+  const [title, setTitle] = useState<string>('');
+  const [userId, setUserId] = useState<number>(0);
+  const [error, setError] = useState<boolean>(false);
+
+  const reset = () => {
+    setTitle('');
+    setUserId(0);
+    setError(false);
+  };
+
+  const addTodo = () => {
+    const id: number =
+      visibleTodos.reduce((acc, to) => {
+        if (to.id > acc) {
+          return to.id;
+        }
+
+        return acc;
+      }, 0) + 1;
+
+    const todo = {
+      id,
+      title,
+      userId,
+      completed: false,
+      user: getUserById(userId),
+    };
+
+    setVisibleTodos(prevTodos => [...prevTodos, todo]);
+  };
+
+  const handleSubmission = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (!title || !userId) {
+      setError(true);
+
+      return;
+    }
+
+    addTodo();
+    reset();
+  };
+
+  const usersToChoose = usersFromServer.map(user => {
+    return {
+      value: user.id,
+      text: user.name,
+    };
+  });
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/todos" method="POST">
+      <form
+        action="/api/todos"
+        onSubmit={event => handleSubmission(event)}
+        method="POST"
+      >
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <input
+            type="text"
+            placeholder="add a title"
+            onChange={event => {
+              setTitle(event.target.value);
+            }}
+            data-cy="titleInput"
+            value={title}
+          />
+          {!title && error && (
+            <span className="error">Please enter a title</span>
+          )}
```

This error handling logic is very close, but has a small flaw. If a user starts typing (which correctly hides the error) and then deletes all the text, this message will reappear before they click 'Add' again. The requirement "hide the message immediately after any change" implies the error should stay hidden until the next submission attempt.

A more robust approach would be to use an object to manage errors for each field separately. For example:
```javascript
const [errors, setErrors] = useState({ title: false, user: false });
```
Then, in your `onChange` handlers, you can clear the error for just the specific field being changed.